File: Base/Desktop/Allors/Client/Base/Angular/Context.ts
namespace Allors {
    export class Context {
        $q: ng.IQService;

        session: ISession;

        objects: { [name: string]: ISessionObject; } = {};
        collections: { [name: string]: ISessionObject[]; } = {};
        values: { [name: string]: any; } = {};

        constructor(
            public name: string,
            public database: Database,
            public workspace: Workspace) {

            this.$q = this.database.$q;

            this.session = new Session(this.workspace);
        }

        load(params?: any): ng.IPromise<any> {
            return this.$q((resolve, reject) => {

                return this.database.sync(this.name, params)
                    .then((response: Data.Response) => {
                        try {
                            const requireLoadIds = this.workspace.diff(response);

                            if (requireLoadIds.objects.length > 0) {
                                this.database.load(requireLoadIds)
                                    .then((loadResponse: Data.LoadResponse) => {
                                        this.workspace.load(loadResponse);
                                        this.update(response);
                                        this.session.reset();
                                        resolve();
                                    })
                                    .catch(e2 => reject(e2));
                            } else {
                                this.update(response);
                                this.session.reset();
                                resolve();
                            }
                        } catch (e) {
                            reject(e);
                        }
                    })
                    .catch(e => reject(e));

            });
        }

        query(service: string, params: any): ng.IPromise<Result> {
            return this.$q((resolve, reject) => {

                this.database.sync(service, params)
                    .then(v => {
                        try {
                            const response = v as Data.Response;
                            const requireLoadIds = this.workspace.diff(response);

                            if (requireLoadIds.objects.length > 0) {
                                this.database.load(requireLoadIds)
                                    .then(u => {
                                        var loadResponse = u as Data.LoadResponse;
                                        this.workspace.load(loadResponse);
                                        const result = new Result(this.session, response);
                                        resolve(result);
                                    })
                                    .catch((e2) => reject(e2));
                            } else {
                                const result = new Result(this.session, response);
                                resolve(result);
                            }
                        } catch (e) {
                            reject(e);
                        }
                    })
                    .catch((e) => reject(e));

            });
        }

        save(): ng.IPromise<Data.SaveResponse> {
            return this.$q((resolve, reject) => {

                try {
                    const saveRequest = this.session.saveRequest();
                    this.database.save(saveRequest)
                        .then((saveResponse: Data.SaveResponse) => {
                            resolve(saveResponse);
                        })
                        .catch((e2) => reject(e2));
                } catch (e) {
                    reject(e);
                }

            });
        }

        invoke(method: Method): ng.IPromise<Data.InvokeResponse>;
        invoke(service: string, args?: any): ng.IPromise<Data.InvokeResponse>;
        invoke(methodOrService: Method | string, args?: any): ng.IPromise<Data.InvokeResponse> {
            if (methodOrService instanceof Method) {
                return this.database.invoke(methodOrService);
            } else {
                return this.database.invoke(methodOrService as string, args);
            }
        }
        
        private update(response: Data.Response): void {

            _.map(response.namedObjects, (v, k) => {
                this.objects[k] = this.session.get(v);
            });

            _.map(response.namedCollections, (v, k) => {
                this.collections[k] = _.map(v, (obj) => { return this.session.get(obj) });
            });

            _.map(response.namedValues, (v, k) => {
                this.values[k] = v;
            });
        }
    }
}

File: Base/Workspace.Domain/Base/Client/Context.cs
namespace Allors.Workspace.Client
{
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;

    using Data;

    public class Context
    {
        private readonly string name;
        private readonly Database database;
        private readonly Workspace workspace;

        public Context(string name, Database database, Workspace workspace)
        {
            this.name = name;
            this.database = database;
            this.workspace = workspace;

            this.Session = new Session(this.workspace);
        }

        public Session Session { get; }

        public Dictionary<string, SessionObject> Objects { get; private set; }

        public Dictionary<string, SessionObject[]> Collections { get; private set; }

        public Dictionary<string, object> Values { get; private set; }

        public async Task Load(object args)
        {
            var response = await this.database.Pull(this.name, args);
            var requireLoadIds = this.workspace.diff(response);
            if (requireLoadIds.objects.Length > 0)
            {
                var loadResponse = await this.database.Sync(requireLoadIds);
                this.workspace.sync(loadResponse);
            }

            this.Update(response);
            this.Session.Reset();
        }

        public async Task<Result> Query(string service, object args)
        {
            var pullResponse = await this.database.Pull(service, args);
            var requireLoadIds = this.workspace.diff(pullResponse);
            if (requireLoadIds.objects.Length > 0)
            {
                var loadResponse = await this.database.Sync(requireLoadIds);
                this.workspace.sync(loadResponse);
            }

            var result = new Result(this.Session, pullResponse);
            return result;
        }

        public async Task<PushResponse> Save()
        {
            var saveRequest = this.Session.PushRequest();
            var pushResponse = await this.database.Push(saveRequest);
            this.Session.PushResponse(pushResponse);
            return pushResponse;
        }

        public Task<InvokeResponse> Invoke(Method method)
        {
            return this.database.Invoke(method);
        }

        public Task<InvokeResponse> Invoke(string service, object args)
        {
            return this.database.Invoke(service, args);
        }
        
        private void Update(PullResponse response)
        {
            this.Objects = response.namedObjects.ToDictionary(
                pair => pair.Key,
                pair => this.Session.Get(long.Parse(pair.Value)));
            this.Collections = response.namedCollections.ToDictionary(
                pair => pair.Key,
                pair => pair.Value.Select(v => this.Session.Get(long.Parse(v))).ToArray());
            this.Values = response.namedValues.ToDictionary(
                pair => pair.Key,
                pair => pair.Value);
        }
    }
}